File: slp/backend-dotnet/Features/User/IUserRepository.cs
using backend_dotnet.Features.Dashboard;

namespace backend_dotnet.Features.User
{
    public interface IUserRepository
    {
        Task<User?> GetByIdAsync(int id);
        Task<User?> GetByEmailAsync(string email);
        Task<User?> GetByUsernameAsync(string username);
        Task<User?> GetByResetTokenAsync(string token);
        Task<User?> GetByEmailVerificationTokenAsync(string token);

        Task UpdateAsync(User user);
        Task CreateAsync(User user);
        Task DeleteAsync(User user);
        Task<(IEnumerable<User> Items, int TotalCount)> GetAllAsync(string? search, int page, int pageSize);
        Task<UserStatsDto> GetUserStatsAsync(int userId);
    }
}

File: slp/backend-java/src/main/java/com/app/backendjava/features/user/IUserRepository.java
package com.app.backendjava.features.user;

import com.app.backendjava.features.dashboard.UserStatsDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.Optional;

public interface IUserRepository {
    Optional<User> getByIdAsync(int id);
    Optional<User> getByEmailAsync(String email);
    Optional<User> getByUsernameAsync(String username);
    Optional<User> getByResetTokenAsync(String token);
    Optional<User> getByEmailVerificationTokenAsync(String token);

    User saveAsync(User user); // Handles both Create and Update
    void deleteAsync(User user);

    Page<User> getAllAsync(String search, Pageable pageable);
    UserStatsDto getUserStatsAsync(int userId);
}